File: backend/src/utils/activityLog.js
/**
 * activityLog.js — structured logging for the four core login/logout
 * flows, distinct from logger.error (failures only), so they're easy to
 * filter for in the combined log or an APM without string-matching emoji.
 */
const logger = require('./logger');

function logDayLogin(username, lat, lng) {
  logger.info('day_login', { username, lat, lng });
}

function logDayLogout(username, durationMins, distanceKm) {
  logger.info('day_logout', { username, durationMins, distanceKm: parseFloat(distanceKm || 0) });
}

function logDealerLogin(username, dealerName) {
  logger.info('dealer_login', { username, dealerName });
}

function logDealerLogout(username, dealerName, durationMins, outOfRadius) {
  logger.info('dealer_logout', { username, dealerName, durationMins, outOfRadius: !!outOfRadius });
}

function logVisitInterrupted(username, dealerName, distanceMeters) {
  logger.warn('visit_interrupted', { username, dealerName, distanceMeters });
}

module.exports = { logDayLogin, logDayLogout, logDealerLogin, logDealerLogout, logVisitInterrupted };

File: backend-py/app/utils/activity_log.py
"""activity_log.py — ports activityLog.js exactly (structured info/warn
logs for the four core login/logout flows)."""
from app.core.logging_config import log_info, log_warn


def log_day_login(username: str, lat, lng) -> None:
    log_info("day_login", username=username, lat=lat, lng=lng)


def log_day_logout(username: str, duration_mins: int, distance_km) -> None:
    log_info("day_logout", username=username, durationMins=duration_mins, distanceKm=float(distance_km or 0))


def log_dealer_login(username: str, dealer_name: str) -> None:
    log_info("dealer_login", username=username, dealerName=dealer_name)


def log_dealer_logout(username: str, dealer_name: str, duration_mins: int, out_of_radius: bool) -> None:
    log_info("dealer_logout", username=username, dealerName=dealer_name, durationMins=duration_mins, outOfRadius=bool(out_of_radius))


def log_visit_interrupted(username: str, dealer_name: str, distance_meters) -> None:
    log_warn("visit_interrupted", username=username, dealerName=dealer_name, distanceMeters=distance_meters)
